src/macros.jl

```diff
@@ -474,6 +474,232 @@ function parse_constraint_head(
     return is_vectorized, parse_code, build_call
 end
 
+"""
+    nonlinear_ifelse(a, x, y)
+
+A function that falls back to `ifelse(a, x, y)`, but when called with JuMP
+variables or expressions, returns a [`GenericNonlinearExpr`](@ref).
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x);
+
+julia> nonlinear_ifelse(true, 1.0, 2.0)
+1.0
+
+julia> nonlinear_ifelse(x, 1.0, 2.0)
+ifelse(x, 1.0, 2.0)
+```
+"""
+nonlinear_ifelse(a, x, y) = ifelse(a, x, y)
+
+"""
+    nonlinear_and(x, y)
+
+A function that falls back to `x && y`, but when called with JuMP variables or
+expressions, returns a [`GenericNonlinearExpr`](@ref).
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x);
+
+julia> nonlinear_and(true, false)
+false
+
+julia> nonlinear_and(true, x)
+true && x
+```
+"""
+nonlinear_and(x, y) = x && y
+
+"""
+    nonlinear_or(x, y)
+
+A function that falls back to `x || y`, but when called with JuMP variables or
+expressions, returns a [`GenericNonlinearExpr`](@ref).
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x);
+
+julia> nonlinear_or(true, false)
+true
+
+julia> nonlinear_or(true, x)
+true || x
+```
+"""
+nonlinear_or(x, y) = x || y
+
+"""
+    nonlinear_less_than(x, y)
+
+A function that falls back to `x < y`, but when called with JuMP variables or
+expressions, returns a [`GenericNonlinearExpr`](@ref).
+
+## Example
+
+```jldoctest
+julia> model = Model();
+
+julia> @variable(model, x);
+
+julia> nonlinear_less_than(1, 2)
+true
+
+julia> nonlinear_less_than(x, 2)
+x < 2
+```
+"""
+nonlinear_less_than(x, y) = x < y
```

Comment:
Does `nonlinear_less_than = UserDefinedFunction(:<, isless)` work ?

Reply:
Yes, but this also works. Note that we'd need to define it as `const NONLINEAR_LESS_THAN = User...`. That makes it a little more clunky to use manually outside the macros. I'd prefer to keep this as-is.

Reply:
Indeed, I didn't think about the naming convention. Maybe we can make an exception for constant that are actually representing functions. As a matter of fact, Julia treats functions as constants so it's not so far off
```julia
julia> f() = 1
f (generic function with 1 method)

julia> f = 2
ERROR: invalid redefinition of constant f
Stacktrace:
 [1] top-level scope
   @ REPL[2]:1
```

Reply:
Why can't we just keep this function as-is? It's more efficient than the `UserDefinedFunction`. We don't need to complicate things.

Reply:
@blegat's suggestion is to rename `UserDefinedFunction` to `AtomFunction` (or ...), and also provide faster fallbacks for the `(::AtomFunction)(args...)` case with one or two arguments.

Reply:
`const atom_less_than = AtomFunction(:<, <)` or `const ATOM_LESS_THAN = AtomFunction(:<, <)`?